### DISCUSSION TITLE
Suggested Speed Values for Waveshare Jetbot kit

### DISCUSSION THREAD

**Discussion by DingoOz** [NVIDIA-AI-IOT/jetbot#353] [General]
I bought a Jetbot kit from Waveshare in November 2020.

After using it with the standard Jetbot code I have found that the motors are geared differently to the "original" Jetbot. The Waveshare motors are generally too fast on the default values.

I would recommend that when moving, set the speed to 0.11 and work up from there.

In the Object Following notebook. I would suggest using the sliders to set the speed to 0.1 and the turn rate (proportional control) to 0.2 for the Waveshare Jetbot.

These values allowed it to track an object successfully, without overshooting the target, or going too fast for the machine vision processing framerate.

**Comment by tokk-nv** [comment]
Hi DingoOz,
Thank you for being the first one to post on this GitHub Discussions!

And thank you for sharing your recommended actual speed gain values for the Waveshare kit.
It sounds like you get to take advantage of the new slider in the Road Following notebook that we introduced recently (https://github.com/NVIDIA-AI-IOT/jetbot/pull/342).

We also today merged a fix on the training notebook for Road Following today.
Together with those improvements, I hope people will have more success getting JetBot to perform the Road Following.

I also find that the Waveshare JetBot motors spin much faster, and I've been suspecting that their motor driver takes VBAT (12V) as the motor driving voltage rather than 5V which is the case of the original open-source JetBot.

**Reply by DingoOz** [reply]
You are correct, I just had a look at the schematic for the Waveshare board and the motors are being driven from a 12V line.

https://www.waveshare.com/w/upload/1/16/JetBot.pdf

![image](https://user-images.githubusercontent.com/1388693/104423770-3ff97580-55ca-11eb-947b-2047ffc696c8.png)

**Reply by dbreggin** [reply]
This is an incomplete thought because I haven't done a deep dive on it yet (so it's all unresearched theory at this point), but one possibility is regulating the voltage into the motor block to 5VDC.  This could be accomplished by pulling from the 5V circuit on the JetBot (there may be current draw vs capacity issues here), or by inserting a new 5VDC regulator circuit upstream of the motor controller board.

The same issue shows up in TeleOp.  The robot is just too twitchy to control while driving.  Now that I am aware of the 12V vs 5V issue, voltage regulation may be a better (or alternative, or additional) solution than I was planning.  The plan is to acquire a pair of motors that are geared lower (80 rpm vs 160 rpm @5VDC) that also have encoders on them.  Then the lower gearing will slow down the wheel's max speed, and use the encoders (rather than the voltage) to control the speed of the motors.

**Reply by DingoOz** [reply]
I think the most simple hardware modification is to use a voltage divider to bring 12v down to 6V. This would simply be cutting the red wire for each motor and added two identical resistors and feeding the motor from 'between' the two resitstors.
https://learn.sparkfun.com/tutorials/voltage-dividers/all#:~:text=A%20voltage%20divider%20is%20a,most%20fundamental%20circuits%20in%20electronics.

The downside of this approach is you waste some power as heat from the resistors themselves (and depending on values, they can get really hot) but it is very easily achieved with two resistors (per motor).

**Reply by dbreggin** [reply]
DingoOz,  Your voltage divider concept is novel and straightforward, but as you point out is is wasteful of energy (and potentially damaging to the robot from the heat generated).  Usually voltage dividers are used in control circuits and not in power circuits like these due to the heat and current issues you have correctly predicted.

After researching the WaveShare board and circuit designs, my original plan of regulating the voltage into the motor driver chips is impractical, as those chips require >6VDC to operate, and there is no obvious way to mod the board to do this. Also on the impractical list is swapping out the battery pack for a 5VDC pack, again due to the >6VDC requirement.  

After researching the kits listed on NVIDIA's partner page, I discovered that ALL of them use a 5VDC battery supply.  That was the tipping point for me.  I am abandoning the WaveShare board altogether, replacing it with the components used by the SparkFun JetBot. This is more complex, but circumvents the VBATT (12VDC) vs 5VDC voltage issue altogether, and I'll end up with a project much closer to the original intended open source design.  I have yet to determine if I will use the original or reduced gear motors.  Other JetBots I have seen with 5VDC batteries and the yellow motors are controllable, but still kind of fast and twitchy, so the geared down (blue) motors are still in the running for the final version.

I found the list of parts I needed on the SparkFun website, and purchased them that afternoon from my local MicroCenter.  The only thing not addressed is the Waveshare board's mechanical function in supporting the Jetson Nano board.  I have chosen a piece of 0.92" clear Lexan (HomeDepot $5) as the replacement material.  Due to the size of the battery, I have also rotated the Jetson orientation by 90 degrees for clearance of the standoffs between the boards.  SparkFun's hardware does require some software adjustments which are well documented in SparkFun's instructions.  Since I'm staying with the WaveShare supplied camera, I'm not sure if the camera differences will make any issues in software or not. I'm also staying with the WaveShare chassis because their design is really solid.  I could 3d print my own, but it wouldn't be near as nice as theirs.

In response to a letter sent to WaveShare: 
_"The motors should be used in 12V.  
We provide schematic of the Jetbot board on wiki, please check it.  
We do not provide layout, I'm sorry"._  

The didn't address the overspeed aspect of their design at all.

I'll post further updates as the project moves forward.

**Reply by chitoku** [reply]
Hi dbreggin, DingOz,
Please check out how I created the 5V using the DC/DC converter.
https://github.com/NVIDIA-AI-IOT/jetbot/discussions/377#discussioncomment-691685

**Comment by DingoOz** [comment]
Additionally, for the Road Following example I have found the following slider values work well for the Waveshare Jetbot:

Speed Gain: 0.11
Steering Gain: 0.03
Steering kd: 0.17 (or a little higher)
Steering bias: 0.00

Note the road I created for my Waveshare Jetbot is only about twice the width of the robot itself so these values help it navigate a quite tight course.
You can see the size and layout of turns [here](https://adventuresinsilicon.blogspot.com/2021/01/cheap-and-easy-robot-racetrack-or-line.html)

**Comment by aixian-jiang** [comment]
Hi, I am totally a newbie and just bought a waveshare jetbot kit with nano 2GB. when I try to download jetbot image from website (http://jetbot.org/master/software_setup/sd_card.html), there is a new version Jetpack 4.5, do I need to download the new version or old version? One more question, Waveshare website suggests to unzip the zip file to .img before we flash it to SD card, while does not mention this here, is that correct?

**Reply by santaimpersonator** [reply]
Once you unzip the downloaded (`.zip`) file, you'll find the image (`.img`) file that you want to flash to the SD card.

For the 2GB board, with Jetpack 4.5 you'll need to boot the board with the plain image before using a custom image (as mentioned in the call out at the bottom of **Step 1**).

**Comment by tjp1940** [comment]
I believe you want to load the latest version of JetPack OS (JetPack4.5).
Then make an image for the jetbot and load it.  Be sure you use the Nano
2GB choice.  This worked for me.  I did not unzip the file. I formatted a
new SD then used Etcher to flash it. This was done on a PC with Windows10.
TJP

**Comment by tjp1940** [comment]
Hi. I am not an expert but when I built my jetbot I downloaded the latest
version of JetPack (JetPack4.5) first, and then the jetbot image , just
to play it safe.  It worked fine. And *no* I did not unzip the file.  I
just formatted a new SD and flashed it with Etcher loaded with JetPack4.5,
then after that
did the same with the jetbot image. (Used a PC with windows10)
TJP

**Reply by aixian-jiang** [reply]
Thank you for reply! I try to download the jetson nano 2GB first to try out the develop kit, it's working, so when setup jetbot software, I am using docker container. 
but now I encountered another problem when I follow the instruction in youtube (https://www.youtube.com/watch?v=MY_Fe7EN6ro&t=29s), I could not pull file from NGC when run coment ./run_from_ngc.sh. after ./script/configure_jetson.sh and ./set_nvidia_runtime.sh, it shows that -bash: ./run_from_ngc.sh: no such file or directory.

I am lost! anyone can help?

**Reply by santaimpersonator** [reply]
@aixian-jiang There is no script with that name in this repo. Where in the video does it mention that (i.e. use a timestamped video link)?

**Comment by tjp1940** [comment]
I did not use the docker container proceedure.  Instead I followed the
jetbot wiki software instructions.  However it could be that the bash files
you are trying to run are in a different directory - just a thought.
